Migrates todo-mvvm-live-kotlin to androidX, LiveData, removes SingleLiveEvent

JoseAlcerreca opened the pull request:
Merges changes from todo-mvp.

SingleLiveEvent gets replaced with Event.

Fixes the Stats UI tests.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepository.kt:
```diff
@@ -58,6 +59,8 @@ class TasksRepository(
             return
         }
 
+        EspressoIdlingResource.increment() // Set app as busy.
```
oh. we're actually modifying production code to test this? :(

JoseAlcerreca replied:
We always have. No need with mock but you have to in prod.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsActivity.kt:
```diff
@@ -33,7 +33,7 @@ import com.example.android.architecture.blueprints.todoapp.util.setupActionBar
 class StatisticsActivity : AppCompatActivity() {
 
-    private lateinit var drawerLayout: DrawerLayout
+    private lateinit var drawerLayout: androidx.drawerlayout.widget.DrawerLayout
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsActivity.kt:
```diff
@@ -67,7 +67,7 @@ class StatisticsActivity : AppCompatActivity() {
 
     private fun setupNavigationDrawer() {
-        drawerLayout = (findViewById<DrawerLayout>(R.id.drawer_layout)).apply {
+        drawerLayout = (findViewById<androidx.drawerlayout.widget.DrawerLayout>(R.id.drawer_layout)).apply {
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsFragment.kt:
```diff
@@ -15,20 +15,18 @@
  * Main UI for the statistics screen.
  */
-class StatisticsFragment : Fragment() {
+class StatisticsFragment : androidx.fragment.app.Fragment() {
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewModel.kt:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017, The Android Open Source Project
+ * Copyright 2016, The Android Open Source Project
```
2016?

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewModel.kt:
```diff
@@ -34,38 +38,66 @@ import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepo
+class StatisticsViewModel(context: Application, private val tasksRepository: TasksRepository) :
+    AndroidViewModel(context) {
+
+    private val _dataLoading = MutableLiveData<Boolean>()
```
why do we have all of these new lines? is this a standard I don't know of?

JoseAlcerreca replied:
We've always done this this way. I don't think we have specific style guidelines on it.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewModel.kt:
```diff
@@ -34,38 +38,66 @@ import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepo
-                error.set(true)
-                numberOfActiveTasks = 0
-                numberOfCompletedTasks = 0
+                _error.setValue(true)
```
you can use `_error.value = true`

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewModel.kt:
```diff
@@ -75,17 +107,30 @@ class StatisticsViewModel(
+        var completed = 0
+        var active = 0
+
+        for (task in tasks) {
```
consider a `foreach`

JoseAlcerreca replied:
tasks.forEach {  task -> is actually longer.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewModel.kt:
```diff
@@ -75,17 +107,30 @@ class StatisticsViewModel(
+        _dataLoading.value = false
     }
-}
+}
```
new line

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt:
```diff
@@ -28,11 +26,12 @@ import android.widget.CheckBox
  * Main UI for the task detail screen.
  */
-class TaskDetailFragment : Fragment() {
+class TaskDetailFragment : androidx.fragment.app.Fragment() {
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
 package com.example.android.architecture.blueprints.todoapp.taskdetail
```
bring back the copyright

JoseAlcerreca replied:
How did that happen?!

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
-        private set
-    val isDataAvailable
-        get() = task.get() != null
+open class TaskDetailViewModel(context: Application, private val tasksRepository: TasksRepository) :
```
does this need to be open? 
If yes, add a comment explaining why

florina-muntenescu replied:
Does this need to be an `AndroidViewModel`?

JoseAlcerreca replied:
Nope, nope.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
+
+    val deleteTaskCommand = MutableLiveData<Event<Any>>()
+
+    private val _snackbarText = MutableLiveData<Event<Int>>()
```
would you mind moving the `_snackbarText` closer to `snackbarMessage`

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
-        task.get()?.let {
-            tasksRepository.deleteTask(it.id)
-            deleteTaskCommand.call()
+        if (_task.value != null) {
```
`_task.value?.apply { ...`

JoseAlcerreca replied:
Refactored the whole class

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
+
+    private val _dataLoading = MutableLiveData<Boolean>()
+
+    val editTaskCommand = MutableLiveData<Event<Any>>()
```
If you just care that something was triggered, consider `Event<Unit>`

JoseAlcerreca replied:
Much better yeah

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
 
     fun setCompleted(completed: Boolean) {
-        if (isDataLoading) {
+        if (_dataLoading.value!!) {
```
can we avoid `!!` in this method?

JoseAlcerreca replied:
Missed all the kotlinfier !!s. Refactored the whole class to avoid all of them.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
-        taskId?.let {
-            isDataLoading = true
-            tasksRepository.getTask(it, this)
+        if (taskId != null) {
```
`taskId?.apply { `?

florina-muntenescu replied:
this was unresolved without change or explanation. so re-opening it.

JoseAlcerreca replied:
If you use ?.let here (not apply!) `taskId` is not smartly cast to non-null.

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt:
```diff
@@ -1,108 +1,111 @@
+        _snackbarText.value = Event(message)
     }
-}
+}
```
new line

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksActivity.kt:
```diff
@@ -15,26 +15,27 @@
 class TasksActivity : AppCompatActivity(), TaskItemNavigator, TasksNavigator {
 
-    private lateinit var drawerLayout: DrawerLayout
+    private lateinit var drawerLayout: androidx.drawerlayout.widget.DrawerLayout
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.kt:
```diff
@@ -17,28 +17,27 @@
  * Display a grid of [Task]s. User can choose to view all, active or completed tasks.
  */
-class TasksFragment : Fragment() {
+class TasksFragment : androidx.fragment.app.Fragment() {
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.kt:
```diff
@@ -1,42 +1,22 @@
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
 package com.example.android.architecture.blueprints.todoapp.tasks
```
bring back the copyright

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.kt:
```diff
@@ -47,32 +27,76 @@ import com.example.android.architecture.blueprints.todoapp.util.EDIT_RESULT_OK
+        get() = _items
+
+    init {
+        // Force use of Application Context.
```
what's this comment about?

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.kt:
```diff
@@ -47,32 +27,76 @@ import com.example.android.architecture.blueprints.todoapp.util.EDIT_RESULT_OK
-        private val tasksRepository: TasksRepository
+    context: Application,
+    private val tasksRepository: TasksRepository
 ) : AndroidViewModel(context) {
```
does this need to be an AndroidViewModel?

JoseAlcerreca replied:
It does, I'm accessing lots of resources

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.kt:
```diff
@@ -84,49 +108,50 @@ class TasksViewModel(
             TasksFilterType.ALL_TASKS -> {
-                setFilter(R.string.label_all, R.string.no_tasks_all,
-                        R.drawable.ic_assignment_turned_in_24dp, true)
+                _currentFilteringLabel.value = context.getString(R.string.label_all)
```
I kinda' like the method used before (maybe with a better name). why was it removed? can we bring it back?

JoseAlcerreca replied:
Bad merge, done

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.kt:
```diff
@@ -137,74 +162,81 @@ class TasksViewModel(
         }
     }
 
+    private fun showSnackbarMessage(message: Int) {
```
is this method used?

JoseAlcerreca replied:
It is

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.kt:
```diff
@@ -137,74 +162,81 @@ class TasksViewModel(
         })
     }
-}
+}
```
new line

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/util/AppCompatActivityExt.kt:
```diff
@@ -39,7 +39,7 @@ const val EDIT_RESULT_OK = Activity.RESULT_FIRST_USER + 3
  * performed by the `fragmentManager`.
  */
-fun AppCompatActivity.replaceFragmentInActivity(fragment: Fragment, frameId: Int) {
+fun AppCompatActivity.replaceFragmentInActivity(fragment: androidx.fragment.app.Fragment, frameId: Int) {
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/util/AppCompatActivityExt.kt:
```diff
@@ -49,7 +49,7 @@ fun AppCompatActivity.replaceFragmentInActivity(fragment: Fragment, frameId: Int
  * performed by the `fragmentManager`.
  */
-fun AppCompatActivity.addFragmentToActivity(fragment: Fragment, tag: String) {
+fun AppCompatActivity.addFragmentToActivity(fragment: androidx.fragment.app.Fragment, tag: String) {
```
import

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/util/AppCompatActivityExt.kt:
```diff
@@ -68,7 +68,7 @@ fun <T : ViewModel> AppCompatActivity.obtainViewModel(viewModelClass: Class<T>)
  * Runs a FragmentTransaction, then calls commit().
  */
-private inline fun FragmentManager.transact(action: FragmentTransaction.() -> Unit) {
+private inline fun androidx.fragment.app.FragmentManager.transact(action: androidx.fragment.app.FragmentTransaction.() -> Unit) {
```
imports

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/util/ViewExt.kt:
```diff
@@ -40,9 +41,12 @@ fun View.showSnackbar(snackbarText: String, timeLength: Int) {
-        snackbarMessageLiveEvent: SingleLiveEvent<Int>, timeLength: Int) {
-    snackbarMessageLiveEvent.observe(lifecycleOwner, Observer {
-        it?.let { showSnackbar(context.getString(it), timeLength) }
+        snackbarEvent: LiveData<Event<Int>>, timeLength: Int) {
```
formatting

florina-muntenescu commented on todoapp/app/src/main/res/layout/addtask_act.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?><!--
-  ~ Copyright (C) 2017 The Android Open Source Project
+  ~ Copyright (C) 2015 The Android Open Source Project
```
2015?

florina-muntenescu commented on todoapp/app/src/main/res/layout/statistics_act.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?><!--
-  ~ Copyright (C) 2017 The Android Open Source Project
+  ~ Copyright (C) 2015 The Android Open Source Project
```
2015?

florina-muntenescu commented on todoapp/app/src/main/res/layout/statistics_frag.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?><!--
-  ~ Copyright (C) 2017 The Android Open Source Project
+  ~ Copyright (C) 2015 The Android Open Source Project
```
2015?

JoseAlcerreca replied:
Yep, it is the year it was created on. Apparently that's what you're supposed to do.

florina-muntenescu commented on todoapp/app/src/main/res/layout/taskdetail_act.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?><!--
-  ~ Copyright (C) 2017 The Android Open Source Project
+  ~ Copyright (C) 2015 The Android Open Source Project
```
2015?

JoseAlcerreca replied:
Yep, it is the year it was created on. Apparently that's what you're supposed to do.

florina-muntenescu commented on todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/LiveDataTestUtil.java:
```diff
@@ -0,0 +1,48 @@
+        //noinspection unchecked
+        return (T) data[0];
+    }
+}
```
new line

florina-muntenescu commented on todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModelTest.kt:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017, The Android Open Source Project
+ * Copyright 2016, The Android Open Source Project
```
2016?

JoseAlcerreca replied:
Yep, it is the year it was created on. Apparently that's what you're supposed to do.

florina-muntenescu commented on todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModelTest.kt:
```diff
@@ -97,147 +96,143 @@ class TasksViewModelTest {
     }
-}
+
+}
```
new line

chrisbanes commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/Event.kt:
```diff
@@ -0,0 +1,27 @@
+package com.example.android.architecture.blueprints.todoapp
```
Copyright

chrisbanes commented on todoapp/app/src/main/AndroidManifest.xml:
```diff
@@ -1,5 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!--
+<?xml version="1.0" encoding="utf-8"?><!--
```
Revert

chrisbanes commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/ScrollChildSwipeRefreshLayout.kt:
```diff
@@ -31,7 +31,7 @@ import android.view.View
         context: Context,
         attrs: AttributeSet? = null
-) : SwipeRefreshLayout(context, attrs) {
+) : androidx.swiperefreshlayout.widget.SwipeRefreshLayout(context, attrs) {
```
Revert

chrisbanes commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt:
```diff
@@ -55,21 +54,22 @@ class AddEditTaskFragment : Fragment() {
-        activity.findViewById<FloatingActionButton>(R.id.fab_edit_task_done).apply {
-            setImageResource(R.drawable.ic_done)
-            setOnClickListener { viewDataBinding.viewmodel?.saveTask() }
+        activity?.findViewById<FloatingActionButton>(R.id.fab_edit_task_done)?.let {
```
Why? apply is right here (or run)

JoseAlcerreca replied:
I prefer not to change receivers, especially in samples. It makes things harder to read.

Murray68w commented:
I'm not sure how I got on this thread but is there any way I can
unsubscribe from these? I get emails on the daily and they need to stop.

On Wed, Feb 6, 2019, 2:36 AM Chris Banes <notifications@github.com> wrote:

> *@chrisbanes* commented on this pull request.
> ------------------------------
>
> In
> todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/Event.kt
> <https://github.com/googlesamples/android-architecture/pull/630#discussion_r254160928>
> :
>
> > @@ -0,0 +1,27 @@
> +package com.example.android.architecture.blueprints.todoapp
>
> Copyright
> ------------------------------
>
> In todoapp/app/src/main/AndroidManifest.xml
> <https://github.com/googlesamples/android-architecture/pull/630#discussion_r254160987>
> :
>
> > @@ -1,5 +1,4 @@
> -<?xml version="1.0" encoding="utf-8"?>
> -<!--
> +<?xml version="1.0" encoding="utf-8"?><!--
>
> Revert
> ------------------------------
>
> In
> todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/ScrollChildSwipeRefreshLayout.kt
> <https://github.com/googlesamples/android-architecture/pull/630#discussion_r254161046>
> :
>
> > @@ -31,7 +31,7 @@ import android.view.View
>  class ScrollChildSwipeRefreshLayout @JvmOverloads constructor(
>          context: Context,
>          attrs: AttributeSet? = null
> -) : SwipeRefreshLayout(context, attrs) {
> +) : androidx.swiperefreshlayout.widget.SwipeRefreshLayout(context, attrs) {
>
> Revert
> ------------------------------
>
> In
> todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt
> <https://github.com/googlesamples/android-architecture/pull/630#discussion_r254161236>
> :
>
> >          setHasOptionsMenu(true)
>          retainInstance = false
>          return viewDataBinding.root
>      }
>
>      private fun setupFab() {
> -        activity.findViewById<FloatingActionButton>(R.id.fab_edit_task_done).apply {
> -            setImageResource(R.drawable.ic_done)
> -            setOnClickListener { viewDataBinding.viewmodel?.saveTask() }
> +        activity?.findViewById<FloatingActionButton>(R.id.fab_edit_task_done)?.let {
>
> Why? apply is right here (or run)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/googlesamples/android-architecture/pull/630#pullrequestreview-200439704>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AZHcdnZOY3sLwwGyvgHctqVN5SWfFQt8ks5vKoYCgaJpZM4ajz0k>
> .
>

JoseAlcerreca commented on todoapp/app/src/androidTest/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksScreenTest.kt:
```diff
@@ -124,19 +147,19 @@ import org.junit.runner.RunWith
 
         // Click on the edit task button
-        onView(withId(R.id.fab_edit_task)).perform(click())
+        onView(withId(id.fab_edit_task)).perform(click())
```
LOL how did this happen

JoseAlcerreca commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.kt:
```diff
@@ -31,89 +15,112 @@ import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepo
         }
     }
 
+    fun getDataLoading(): LiveData<Boolean> {
```
Missed this one!

JoseAlcerreca commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.kt:
```diff
@@ -31,89 +15,112 @@ import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepo
-
-    private fun showSnackbarMessage(@StringRes message: Int) {
-        snackbarMessage.value = message
+        _taskUpdated.value = Event(Any())
```
Use Unit everywhere instead of Any.

JoseAlcerreca commented:
@Murray68w just unwatch the repo at https://github.com/googlesamples/android-architecture

I moved all the livedata getters close to the backing field and addressed the comments. Good catches @florina-muntenescu and @chrisbanes .

florina-muntenescu commented on todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.kt:
```diff
@@ -1,27 +1,11 @@
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
 package com.example.android.architecture.blueprints.todoapp.addedittask
```
copyright

JoseAlcerreca replied:
Done!

JoseAlcerreca commented:
Comments addressed

JoseAlcerreca commented:
Circle is finally green, merging.
